src/main/groovy/io/openliberty/tools/gradle/tasks/GenerateFeaturesTask.groovy

```diff
@@ -0,0 +1,348 @@
+            logger.debug("Exception creating server.xml object model", e);
+        }
+        return null;
+    }
+
+    /**
+     * Add a comment to server.xml to warn them we created another file with features in it.
+     */
+    private void addGenerationCommentToConfig(ServerConfigXmlDocument doc, File serverXml) {
+        if (doc == null) {
+            return;
+        }
+        try {
+            if (doc.createFMComment(FEATURES_FILE_MESSAGE)) {
+                doc.writeXMLDocument(serverXml);
+                XmlDocument.addNewlineBeforeFirstElement(serverXml);
+            }
+        } catch (IOException | TransformerException e) {
+            log.debug("Exception adding comment to server.xml", e);
+        }
+        return;
+    }
+
+    private Set<String> getClassesDirectories() {
+        Set<String> classesDirectories = new ArrayList<String>();
+        project.sourceSets.main.getOutput().getClassesDirs().each {
+            classesDirectories.add(it.getAbsolutePath());
+        }
+        return classesDirectories;
+    }
+
+    private getEEVersion(Object project) {
+        String eeVersion = null
+        project.configurations.compile.allDependencies.each {
+            dependency ->
+                if (dependency.group.equals("io.openliberty.features")) {
+                    if (dependency.name.equals("javaee-7.0")) {
+                        eeVersion = "ee7"
+                    } else if (dependency.name.equals("javaee-8.0")) {
+                        eeVersion = "ee8"
+                    } else if (dependency.name.equals("javaeeClient-7.0")) {
+                        eeVersion = "ee7"
+                    } else if (dependency.name.equals("javaeeClient-8.0")) {
+                        eeVersion = "ee8"
+                    } else if (dependency.name.equals("jakartaee-8.0")) {
+                        eeVersion = "ee8"
+                    }
+                } else if (dependency.group.equals("jakarta.platform") &&
+                        dependency.name.equals("jakarta.jakartaee-api") &&
+                        dependency.version.equals("8.0.0")) {
+                    eeVersion = "ee8";
+                }
+        }
+        return eeVersion;
+    }
+
+    private getMPVersion(Object project) {
+        String mpVersion = null
+        project.configurations.compile.allDependencies.each {
+            if (it.group.equals("org.eclipse.microprofile") &&
+                    it.name.equals("microprofile")) {
+                if (it.version.startsWith("1")) {
+                    mpVersion = "mp1"
+                } else if (it.version.startsWith("2")) {
+                    mpVersion = "mp2"
+                } else if (it.version.startsWith("3")) {
+                    mpVersion = "mp3"
+                } else if (it.version.startsWith("4")) {
+                    mpVersion = "mp4"
+                }
+            }
+        }
+        return mpVersion;
+    }
```

Comment:
Some of these methods seem to be exact duplicates in Maven and Gradle. Would be better to have it in common eventually.

Reply:
Opened https://github.com/OpenLiberty/ci.maven/issues/1342 to deal with this after snapshot release.